### PR TITLE
LibGfx/TIFF: Add support for images with CCITT3 1D compression

### PR DESCRIPTION
**Pull request by LucasChollet**
This compression (tag Compression=2) is not very popular on its own, but a base to implement CCITT3 2D and CCITT4 compressions.

As the format has no real benefits, it is quite hard to find an app that accepts tho encode that for you. So I used the following program that calls `libtiff` directly:
```cpp
#include <vector>
#include <cstdlib>
#include <iostream>

#include <tiffio.h>

// An array containing 0 and 1 of length width * height.
extern std::vector<uint8_t> array;
int main() {
    // From: https://stackoverflow.com/a/34257789

    TIFF *image = TIFFOpen("input.tif", "w");
    int const width = 400;
    int const height = 300;
    TIFFSetField(image, TIFFTAG_IMAGEWIDTH, width);
    TIFFSetField(image, TIFFTAG_IMAGELENGTH, height);
    TIFFSetField(image, TIFFTAG_PHOTOMETRIC, 0);
    TIFFSetField(image, TIFFTAG_COMPRESSION, COMPRESSION_CCITTRLE);

    TIFFSetField(image, TIFFTAG_BITSPERSAMPLE, 1);
    TIFFSetField(image, TIFFTAG_SAMPLESPERPIXEL, 1);
    TIFFSetField(image, TIFFTAG_ROWSPERSTRIP, 1);

    std::vector<uint8_t> scan_line(width / 8 + 8, 0);
    int count = 0;
    for (int i = 0; i < height; i++) {
        std::fill(scan_line.begin(), scan_line.end(), 0);
        for (int x = 0; x < width; ++x) {
            uint8_t eight_pixels = scan_line.at(x / 8);
            eight_pixels = eight_pixels << 1;
            eight_pixels |= !array.at(i * width + x);
            scan_line.at(x / 8) = eight_pixels;
        }
        int bytes = int(width / 8.0 + 0.5);
        if (TIFFWriteScanline(image, scan_line.data(), i, bytes) != 1)
            std::cerr << "Something went wrong\n";
    }

    TIFFClose(image);
}
```

### CODE REVIEW

File: Userland/Libraries/LibGfx/ImageFormats/CCITTDecoder.cpp [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+
+namespace {
+
+struct Code {
+    u16 run_length {};
+    u8 code_length {};
+    u16 code {};
+};
+
+constexpr Array white_terminating_codes = {
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/CCITTDecoder.cpp:23]: Can we get some spec comments here? Official spec is probably one of these: https://www.itu.int/rec/T-REC-T/en

https://www.fileformat.info/mirror/egff/ch09_05.htm is a good overview and probably also worth linking to.

File: Userland/Libraries/LibGfx/ImageFormats/CCITTDecoder.h [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+// You can find a great overview of CCITT compression schemes here:
+// https://www.fileformat.info/mirror/egff/ch09_05.htm
+
+// The CCITT3 specification is accessible at this page:
+// https://www.itu.int/rec/T-REC-T.4/en
+
+// The unidimensional scheme is originally described in:
+// 4.1 One-dimensional coding scheme
+// However, this function implement the TIFF variant (see TIFFLoader.h for a spec link),
```
Comment (nico) [inline Userland/Libraries/LibGfx/ImageFormats/CCITTDecoder.h:21]: s/implement/implements/

**Review by nico — APPROVED** [review]
This is great.

One reason I'm excited about this is that CCITT is one of the missing PDF filters, and hopefully LibPDF will be able to call into this code at some point.

Section 3.3.5 CCITTFaxDecode Filter in the PDF 1.7 spec describes how things look over there. A short summary:

"""
Key K: 
A code identifying the encoding scheme used:
* <0 Pure two-dimensional encoding (Group 4)
* 0 Pure one-dimensional encoding (Group 3, 1-D)
* >0 Mixed one- and two-dimensional encoding (Group 3, 2-D), in which a line encoded one-dimensionally can be followed by at most K − 1 lines encoded two-dimensionally

Key EndOfLine
A flag indicating whether end-of-line bit patterns are required to be present in the encoding. The CCITTFaxDecode filter always accepts end-of-line bit patterns, but requires them only if EndOfLine is true. Default value: false.

Key EncodedByteAlign
A flag indicating whether the filter expects extra 0 bits before each encoded line so that the line begins on a byte boundary. If true, the filter skips over encoded bits to begin decoding each line at a byte boundary. If false, the filter does not expect extra bits in the encoded representation. Default value: false.

Key EndOfBlock
A flag indicating whether the filter expects the encoded data to be terminated by an end-of-block pattern, overriding the Rows parameter. If false, the filter stops when it has decoded the number of lines indicated by Rows or when its data has been exhausted, whichever occurs first. The end-of-block pattern is the CCITT end-of-facsimile-block (EOFB) or return-to-control (RTC) appropriate for the K parameter. Default value: true.

Key BlackIs1
A flag indicating whether 1 bits are to be interpreted as black pixels and 0 bits as white pixels, the reverse of the normal PDF convention for image data. Default value: false.

Key DamagedRowsBeforeError
The number of damaged rows of data to be tolerated before an error occurs. This entry applies only if EndOfLine is true and K is non-negative. Tolerating a damaged row means locating its end in the encoded data by searching for an EndOfLine pattern and then substituting decoded data from the previous row if the previous row was not damaged, or a white scan line if the previous row was also damaged. Default value: 0.
"""

(There's also width and height.)

So if the API could in time support all these parameters (as an options struct, I imagine), that'd be fantastic.

**Comment by LucasChollet** [comment]
This is actually my second try at this in order to get something compatible with `LibPDF`'s interface. But I don't plan to have full-compatibility on the first iteration. I will implement everything as described by the TIFF spec, and then add any remaining parameter required by PDFs.

**Pull request merged**